2089: Add info about new available versions to the `weave status` report

brb opened the pull request:
The `AvailableVersion` field denotes the latest available version which is retrieved via go-checkpoint.
If checkpoint is disabled, the field will contain the `<checkpoint disabled>` value.

E.g.:
```bash
ubuntu@host:~$ weave status

         Version: git-b7d49cdeb7db
AvailableVersion: 1.4.4
<..>
```
Fixes #2089

awh commented:
```
Version: 1.4 (update check disabled)
Version: 1.4 (update check every 6 hours)
Version: 1.4 (version 1.5 available)
```

rade commented:
Suggested output
```
Version: 1.4.4 (version check disabled)
Version: 1.4.4 (available: 1.4.4 - up to date)
Version: 1.4.4 (available: 1.4.6 - please upgrade!)
```

rade commented:
Docs need updating. In particular troubleshooting.md#weave-status and the README.md

bboreham commented on prog/weaver/http.go:
```diff
@@ -218,16 +220,33 @@ var dnsEntriesTemplate = defTemplate("dnsEntries", `\
+		return "version check update disabled"
+	}
+	if vsn := availableVersion.Load().(string); vsn != "" {
+		return fmt.Sprintf("version %s available - please upgrade!")
```
missing parameter

brb commented:
Updated.

Please merge https://github.com/weaveworks/go-checkpoint/pull/9 first. Afterwards, I need to change the commit id of the go-checkpoint git submodule and then this PR can be merged.

Q: What's about `weave report`? I've added `VersionCheckInfo` field to it which contains the same info as it is shown in `weave status`, but I'd prefer something more computer friendly. E.g.:

* `.VersionMeta.CheckEnabled :: boolean`
* `.VersionMeta.NewVersion :: string`
* `.VersionMeta.NextCheck :: string`

brb commented:
Added `.VersionMeta` to `weave report`.

rade commented on prog/weaver/http.go:
```diff
@@ -114,6 +116,21 @@ var rootTemplate = template.New("root").Funcs(map[string]interface{}{
 		return "disabled"
 	},
 	"trimSuffix": strings.TrimSuffix,
+	"versionMeta": func(meta *VersionMeta) string {
```
could that simply be the `String()` method on `versionMeta`?

brb replied:
Yep, thanks.

rade commented on prog/weaver/http.go:
```diff
@@ -217,17 +234,52 @@ var dnsEntriesTemplate = defTemplate("dnsEntries", `\
-	IPAM    *ipam.Status               `json:"IPAM,omitempty"`
-	DNS     *nameserver.Status         `json:"DNS,omitempty"`
+	Version     string
+	VersionMeta *VersionMeta               `json:"Version,omitempty"`
```
won't that end up with two `Version` keys in the json, i.e. one for the `Version` field and one for the `VersionMeta` field (which you've mapped to the 'Version' field)?

brb replied:
Typo, thanks. I'm adding tests for this PR.

rade commented on prog/weaver/http.go:
```diff
@@ -217,17 +234,52 @@ var dnsEntriesTemplate = defTemplate("dnsEntries", `\
 
 var ipamTemplate = defTemplate("ipamTemplate", `{{printIPAMRanges .Router .IPAM}}`)
 
+type VersionMeta struct {
```
`VersionCheck`?

brb replied:
Looks better. I'm replacing all `[vV]ersionMeta` occurrences by `[vV]ersionCheck`.

rade commented on prog/weaver/main.go:
```diff
@@ -68,7 +76,7 @@ func checkForUpdates() {
 	resp, err := checkpoint.Check(&params)
 	handleResponse(resp, err)
-	checkpoint.CheckInterval(&params, updateCheckPeriod, handleResponse)
+	checker = checkpoint.CheckInterval(&params, updateCheckPeriod, handleResponse)
```
As discussed, we should change `CheckInterval` s.t. it performs a check immediately. That way we don't need the `Check` call, can get rid of the intermediate vars for `handleResponse` and `params`, and, most importantly, don't need to run this function in its own goroutine, thus ensuring that `checker` will always be set, which in turn will simplify the status reporting code.

brb replied:
ACK.

rade commented on prog/weaver/http.go:
```diff
@@ -114,6 +116,21 @@ var rootTemplate = template.New("root").Funcs(map[string]interface{}{
+	"versionMeta": func(meta *VersionMeta) string {
+		if meta == nil {
+			return ""
+		}
```
I don't think it can ever be nil.

brb replied:
Theoretically, it can. E.g., when that goroutine https://github.com/weaveworks/weave/blob/issues/2089-vsn-check/prog/weaver/main.go#L194 gets scheduled after the first `weave status|report` request.

rade replied:
but that cannot happen after the other changes we discussed.

brb replied:
Yes, after changing the `CheckInterval` function, it cannot happen.

brb commented:
Updated.

brb commented:
@rade ready for review.

bboreham commented:
Looks like it needs to be rebased to resolve conflicts on master

bboreham commented on test/config.sh:
```diff
@@ -44,6 +44,8 @@ CHECK_ETHWE_MISSING="test ! -d /sys/class/net/ethwe"
 
 DOCKER_PORT=2375
 
+CHECKPOINT_DISABLE=true
```
does this deserve its own variable?

bboreham replied:
oh, I get it - you override it in test 700

brb replied:
Yep.

bboreham commented on test/config.sh:
```diff
@@ -118,7 +120,7 @@ weave_on() {
     shift 1
     [ -z "$DEBUG" ] || greyly echo "Weave on $host:$DOCKER_PORT: $@" >&2
-    CHECKPOINT_DISABLE=true DOCKER_HOST=tcp://$host:$DOCKER_PORT $WEAVE "$@"
+    CHECKPOINT_DISABLE="$CHECKPOINT_DISABLE" DOCKER_HOST=tcp://$host:$DOCKER_PORT $WEAVE "$@"
```
I think this deserves a comment; presumably you are aiming at the `weave launch` commands?

bboreham replied:
There is at least one `weave launch` that doesn't go through this path - in test 140.

brb replied:
This change does not affect the 140 test. The `CHECKPOINT_DISABLE` thing was present prior this PR.

brb replied:
https://github.com/weaveworks/weave/pull/1954 has introduced `CHECKPOINT_DISABLE` in `weave_on`.

bboreham commented:
Apart from comments LGTM

brb commented:
Rebased.

rade commented on prog/weaver/http.go:
```diff
@@ -116,6 +118,19 @@ var rootTemplate = template.New("root").Funcs(map[string]interface{}{
+			v.NextCheckAt.Format("2006/01/02 15:04:05"))
+	}
+}
+
```
This is out of place. Should come just after the `versionCheck()` function.

brb replied:
Fixed, thanks.

rade commented on prog/weaver/http.go:
```diff
@@ -221,17 +236,38 @@ var dnsEntriesTemplate = defTemplate("dnsEntries", `\
+	DNS          *nameserver.Status         `json:"DNS,omitempty"`
+}
+
+func versionCheck() *VersionCheck {
```
this function should come right after the `VersionCheck` type def.
